### DISCUSSION TITLE
Custom Placeholder Plugin with XML-structure like output

### DISCUSSION THREAD

**Discussion by n-xt** [ueberdosis/tiptap#2473] [General Questions]
I plan an extension where the user can insert placeholders (E.g. "Enter your name") within the text. My goal is that I have a template document afterwards where a user can click on the placeholders and fills out the fields. 

My placeholder consits of different properties (id, type, title, description, value,...). It works well when I add all the properties as attributes, however, I may want more complicated text in the properties (e.g. HTML Code in the description, or some placeholder may be dropdowns and I need to store the options). Hence, storing this data as an attribute isnt sufficient. 

I would prefer a rendered HTML output similar to this

```
<placeholder id="my-plcaeholder-id" type="text">
  <title>Enter your name</title>
  <comment>Please fill in here your first name</comment>
  <value>Martin</value>
</placeholder>
```

I use VueNodeViewRenderer which renders a clickable inline DIV - when the user clicks on it, it opens a poup where the user can define all the properties.

However, what is the next approach. I only want to show the title in the editor, not the other properties. 

Here my code until now:

```
export const TipTapPlaceholder = Node.create({
  name: "placeholder",
  draggable: true,
  group: "inline",
  inline: true,
  content: "inline*",

  addCommands() {
    return {
      addPlaceholder:
        (placeholderAttrs) =>
        ({ chain }) => {
          return chain()
            .focus()

            .insertContent({
              type: "placeholder",
              attrs: {
                "data-type": placeholderAttrs.type,
                "data-id": placeholderAttrs.id ? placeholderAttrs.id : uuidv4(),
              },
              content: [
                {
                  type: "text",
                  text: placeholderAttrs.title,
                },
              ],
            })
            .run();
        },
    };
  },

  parseHTML() {
    return [
      {
        tag: "placeholder",
      },
    ];
  },

  renderHTML({ HTMLAttributes }) {
    return ["placeholder", mergeAttributes(HTMLAttributes), 0];
  },

  addNodeView() {
    return VueNodeViewRenderer(TextEditorPlaceholderElement);
  },
  
  addAttributes() {
    return {
      "data-id": {
        default: uuidv4(),
      },
      "data-type": {
        default: "text",
      },
    };
  },
});
```

**Comment by hanspagel** [comment]
Did you try storing your data as JSON in the attribute?

JSON.stringify
JSON.parse

**Reply by n-xt** [reply]
ok, I think this is the easiest solution. I will go for that for now :) thanks
